### PR TITLE
Throw SDE in the case of a choice with no branches

### PR DESCRIPTION
**Pull request by rsteinberger**
Section 15 of the spec satates that a choice that declares no branches
in the DFDL schema is a schema definition error.

Daffodil-1032

### CODE REVIEW

File: daffodil-core/src/main/scala/org/apache/daffodil/dsom/ChoiceGroup.scala [hunk 1]
```diff
@@ -201,6 +202,12 @@ abstract class ChoiceTermBase(
     }
   }.value
 
+  final lazy val noBranchesFound = LV('noBranchesFound) {
+    if(groupMembers.size.equals(0)) {
```
Comment (stevedlawrence) [inline daffodil-core/src/main/scala/org/apache/daffodil/dsom/ChoiceGroup.scala:206]: The standard in scala is to use ``==`` instead of ``.equals()``. Also note that there should be a space between ``if`` and the open paren.

File: daffodil-test/src/test/scala/org/apache/daffodil/section15/choice_groups/TestChoice.scala [hunk 2]
```diff
@@ -123,9 +124,6 @@ class TestChoice {
   @Test def test_direct_dispatch_16(): Unit = { runnerCH.runOneTest("direct_dispatch_16") }
   @Test def test_direct_dispatch_17(): Unit = { runnerCH.runOneTest("direct_dispatch_17") }
 
-  //@Test def test_choice_noBranch() { runnerCH.runOneTest("choice_noBranch") } - Test consumes no data, which causes a TDMLError
-
-
```
Comment (stevedlawrence) [inline daffodil-test/src/test/scala/org/apache/daffodil/section15/choice_groups/TestChoice.scala:128]: Why did you delete the choice_noBranch test and replace it with a new one? Was something wrong with choice_noBranch? That name feels more appropriate than optionalChoice00, since this isn't testing optional choices.

Comment (rsteinberger) [reply daffodil-test/src/test/scala/org/apache/daffodil/section15/choice_groups/TestChoice.scala:128]: I didn't see the point of using the external xsd. I can changes the name if you like.

Comment (rsteinberger) [reply daffodil-test/src/test/scala/org/apache/daffodil/section15/choice_groups/TestChoice.scala:128]: I renamed it and move it to the basic model.

File: daffodil-core/src/main/scala/org/apache/daffodil/dsom/ChoiceGroup.scala (hunk 1 above)
Comment (rsteinberger) [reply daffodil-core/src/main/scala/org/apache/daffodil/dsom/ChoiceGroup.scala]: Done

File: daffodil-test/src/test/scala/org/apache/daffodil/section15/choice_groups/TestChoice.scala (hunk 2 above)
Comment (stevedlawrence) [reply daffodil-test/src/test/scala/org/apache/daffodil/section15/choice_groups/TestChoice.scala:128]: Gotcha. I suspect this was originally split out beause it's not a valid schema and in some cases that can cause issues. But since it doesn't, seems reasonable to put it along side the other schemas. This change is fine with me.

File: daffodil-test/src/test/scala/org/apache/daffodil/section15/choice_groups/TestChoice.scala [hunk 3]
```diff
@@ -38,6 +38,9 @@ class TestChoice {
 
   import TestChoice._
 
+  @Test def test_choice_noBranch (): Unit = { runnerCH.runOneTest("choice_noBranch") }
+
+  @Test def test_optionalChoice00(): Unit = { runnerCH.runOneTest("optionalChoice00") }
```
Comment (tuxji) [inline daffodil-test/src/test/scala/org/apache/daffodil/section15/choice_groups/TestChoice.scala:43]: Rick, you need to delete test_optionalChoice00 since the "optionalChoice00" test doesn't exist.  This is the reason why the PR checks are failing with the message

```
[error] Test org.apache.daffodil.section15.choice_groups.TestChoice.test_optionalChoice00 failed: org.apache.daffodil.tdml.TDMLExceptionImpl: test optionalChoice00 was not found., took 0.001 sec
[error]     at org.apache.daffodil.tdml.TDMLException$.apply(TDMLException.scala:31)
[error]     at org.apache.daffodil.tdml.DFDLTestSuite.runOneTest(TDMLRunner.scala:388)
[error]     at org.apache.daffodil.tdml.Runner.runOneTest(RunnerFactory.scala:144)
[error]     at org.apache.daffodil.tdml.Runner.runOneTest(RunnerFactory.scala:150)
[error]     at org.apache.daffodil.section15.choice_groups.TestChoice.test_optionalChoice00(TestChoice.scala:43)
[error]     ...
```

**Review by tuxji — COMMENTED** [review]
Other than the failing test, the PR looks good.

File: daffodil-test/src/test/scala/org/apache/daffodil/section15/choice_groups/TestChoice.scala (hunk 3 above)
Comment (rsteinberger) [reply daffodil-test/src/test/scala/org/apache/daffodil/section15/choice_groups/TestChoice.scala]: Yep, thanks John.

**Review by tuxji — APPROVED** [review]
+1

Looks good to me.

**Review by stevedlawrence — APPROVED** [review]
+1 :+1: Thanks for the contribution!
